File: 백준/Gold/2638. 치즈/치즈.py
import sys
input=sys.stdin.readline

def count():
  time=0
  while len(cheese):
    time+=1
    visited=[[False for i in range(M)]for j in range(N)]
    cnt=[[0 for i in range(M)]for j in range(N)]
    melt={}

    q=[[0,0]]
    while q:
      x,y=q.pop()
      if visited[x][y]:
        continue
      visited[x][y]=True
      for i in range(4):
        newX=x+xbox[i]
        newY=y+ybox[i]
        if 0<=newX<N and 0<=newY<M:
          if newX in cheese and newY in cheese[newX]:
            cnt[newX][newY]+=1
            if cnt[newX][newY]>=2:
              if newX in melt:
                melt[newX][newY]=True
              else:
                melt[newX]={newY:True}
          else:
            q.append([newX,newY])

    for x in melt:
      for y in melt[x]:
        cheese[x].pop(y)
        if not cheese[x]:
          cheese.pop(x)

  print(time)


N,M=map(int,input().split())
cheese={}
for i in range(N):
  now=list(map(int,input().split()))
  for j in range(M):
    if now[j]==1:
      if i in cheese:
        cheese[i][j]=True
      else:
        cheese[i]={j:True}

xbox=[-1,1,0,0]
ybox=[0,0,-1,1]
count()

# 8 9
# 0 0 0 0 0 0 0 0 0
# 0 0 0 0 0 0 0 0 0
# 0 1 1 0 0 0 1 1 0
# 0 1 0 1 1 1 0 1 0
# 0 1 0 0 1 0 0 1 0
# 0 1 0 1 1 1 0 1 0
# 0 1 1 0 0 0 1 1 0
# 0 0 0 0 0 0 0 0 0

File: 백준/Gold/2636. 치즈/치즈.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const [X, Y] = input.shift().split(' ').map(Number);
  let l = [];
  let prev = 0;
  for (let i = 0; i < X; i++) {
    const now = input.shift().split(' ').map(Number);
    for (let j = 0; j < Y; j++) {
      if (now[j] == 1) prev += 1;
    }
    l.push(now);
  }

  const xbox = [-1, 1, 0, 0];
  const ybox = [0, 0, -1, 1];

  let left = prev;
  let time = 0;
  while (true) {
    if (left == 0) {
      console.log(time);
      console.log(prev);
      break;
    }
    time += 1;

    let visited = [];
    for (let i = 0; i < X; i++) {
      visited.push([]);
      for (let j = 0; j < Y; j++) {
        visited[i].push(false);
      }
    }
    let q = [[0, 0]];
    visited[0][0] = true;

    prev = 0;
    while (q.length > 0) {
      const [x, y] = q.shift();
      if (l[x][y] == 1) {
        l[x][y] = 0;
        prev += 1;
        continue;
      } else {
        for (let i = 0; i < 4; i++) {
          let newX = x + xbox[i];
          let newY = y + ybox[i];
          if (
            0 <= newX &&
            newX < X &&
            0 <= newY &&
            newY < Y &&
            !visited[newX][newY]
          ) {
            visited[newX][newY] = true;
            q.push([newX, newY]);
          }
        }
      }
    }

    left -= prev;
  }
  process.exit();
});
